File: src/tokamak/processor.py
"""Per-span processing agents.

A Processor takes a reasoning span plus its surrounding context (problem +
answer) and rewrites the reasoning. Two modes are supported:

- compress  — terse rewrite, every step preserved, ~50% fewer tokens
- invert    — expand a compressed skeleton into a fuller trace, no new
              methods or assumptions, conclusion must match the given answer

Modes can be chained (compress → invert) via the pipeline; this module is
single-step.
"""

from __future__ import annotations

import logging
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from typing import List, Optional

from . import prompts
from .caveman import (
    CompressResult,
    compress_rules,
    estimate_tokens,
)
from .llm_client import LLMClient

logger = logging.getLogger("tokamak.processor")


# ---------------------------------------------------------------------------
# Input/output container
# ---------------------------------------------------------------------------

@dataclass
class SpanInput:
    """A reasoning span plus the surrounding rails the processor uses for context."""
    problem: str
    answer: str
    reasoning: str


@dataclass
class ProcessResult:
    mode: str                        # "compress" | "invert" | "noop"
    original: str
    processed: str
    original_tokens: int
    processed_tokens: int

    @property
    def ratio(self) -> float:
        if self.original_tokens == 0:
            return 1.0
        return self.processed_tokens / self.original_tokens


def _format_compress_user(span: SpanInput) -> str:
    return (
        "PROBLEM:\n"
        f"{span.problem.strip() or '(no surrounding problem context provided)'}\n\n"
        "ANSWER:\n"
        f"{span.answer.strip() or '(no final answer provided)'}\n\n"
        "REASONING:\n"
        f"{span.reasoning}"
    )


def _format_invert_user(span: SpanInput) -> str:
    return (
        "Problem:\n"
        f"{span.problem.strip() or '(no surrounding problem context provided)'}\n\n"
        "Model's final answer:\n"
        f"{span.answer.strip() or '(no final answer provided)'}\n\n"
        "Reasoning Bubbles:\n"
        f"{span.reasoning}\n\n"
        "Reconstruct the full reasoning trace."
    )


# ---------------------------------------------------------------------------
# Processor
# ---------------------------------------------------------------------------

class Processor:
    """Run a single processing agent over a span.

    `mode`:
        rules    — pure regex compression of the reasoning text (no LLM,
                   ignores PROBLEM / ANSWER context)
        compress — terse LLM rewrite at intensity `level` (lite|full)
        invert   — LLM expansion of a compressed skeleton
        noop     — passthrough (used for baselines)
    """

    def __init__(
        self,
        mode: str = "compress",
        *,
        level: str = "lite",
        model: Optional[str] = None,
        client: Optional[LLMClient] = None,
    ) -> None:
        if mode not in ("rules", "compress", "invert", "noop"):
            raise ValueError(f"unknown processor mode: {mode!r}")
        self.mode = mode
        self.level = level
        self.model = model
        # rules and noop never touch the network — only build a client if needed.
        self._client = client if client is not None else (
            LLMClient(model=model) if mode in ("compress", "invert") else None
        )

    # ---- single-span ----

    def process(self, span: SpanInput) -> ProcessResult:
        text = span.reasoning
        if not text or not text.strip():
            return ProcessResult(self.mode, text, text, 0, 0)

        if self.mode == "noop":
            return ProcessResult(
                self.mode, text, text,
                estimate_tokens(text), estimate_tokens(text),
            )

        if self.mode == "rules":
            r = compress_rules(text, level=self.level)
            return ProcessResult(
                "rules", r.original, r.compressed,
                r.original_tokens, r.compressed_tokens,
            )

        if self.mode == "compress":
            system = prompts.compress_prompt(self.level)
            user = _format_compress_user(span)
        else:  # invert
            system = prompts.invert_prompt()
            user = _format_invert_user(span)

        try:
            assert self._client is not None
            out = self._client.chat(system=system, user=user)
        except Exception as exc:  # noqa: BLE001
            logger.warning(
                "%s agent failed (%d chars in); falling back to original: %s",
                self.mode, len(text), exc,
            )
            out = text

        return ProcessResult(
            self.mode,
            original=text,
            processed=out,
            original_tokens=estimate_tokens(text),
            processed_tokens=estimate_tokens(out),
        )

    def __call__(self, span: SpanInput) -> ProcessResult:
        return self.process(span)

    # ---- batched ----

    def process_batch(
        self, spans: List[SpanInput], max_workers: int = 8
    ) -> List[ProcessResult]:
        """Process many spans concurrently. Order preserved.

        When the `tokamak_engine` Rust orchestrator is available AND the
        client is configured against an OpenAI-compatible endpoint, the
        whole batch is dispatched through tokio with semaphore-bounded
        concurrency. Otherwise a Python ThreadPoolExecutor parallelizes
        per-span calls.

        Set max_workers close to the endpoint's scheduler width
        (e.g. vLLM `--max-num-seqs`).
        """
        if not spans:
            return []
        if self.mode in ("rules", "noop"):
            return [self.process(s) for s in spans]

        assert self._client is not None
        if self.mode == "compress":
            system = prompts.compress_prompt(self.level)
            format_user = _format_compress_user
        else:
            system = prompts.invert_prompt()
            format_user = _format_invert_user

        items = [{"system": system, "user": format_user(s)} for s in spans]

        try:
            outs = self._client.batch_chat(items, max_workers=max(1, max_workers))
        except Exception as exc:  # noqa: BLE001
            logger.warning(
                "batch %s failed (%d spans); falling back to per-span: %s",
                self.mode, len(spans), exc,
            )
            return [self.process(s) for s in spans]

        results: List[ProcessResult] = []
        for span, out in zip(spans, outs):
            text = span.reasoning
            # Empty out = call failed (Rust path) → fall back to original.
            processed = out if (out and out.strip()) else text
            results.append(ProcessResult(
                mode=self.mode,
                original=text,
                processed=processed,
                original_tokens=estimate_tokens(text),
                processed_tokens=estimate_tokens(processed),
            ))
        return results

File: rust/tokamak/src/processor.rs
//! Per-span processing agent. Compress / invert / rules / noop.

use crate::llm::LlmClient;
use crate::prompts::{self, Level};
use crate::terse::{compress_rules, estimate_tokens};

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Mode { Rules, Compress, Invert, Noop }

impl Mode {
    pub fn from_str(s: &str) -> anyhow::Result<Self> {
        match s {
            "rules"    => Ok(Self::Rules),
            "compress" => Ok(Self::Compress),
            "invert"   => Ok(Self::Invert),
            "noop"     => Ok(Self::Noop),
            other      => anyhow::bail!("unknown processor mode: {other:?}"),
        }
    }
    pub fn name(self) -> &'static str {
        match self {
            Self::Rules => "rules", Self::Compress => "compress",
            Self::Invert => "invert", Self::Noop => "noop",
        }
    }
}

#[derive(Debug, Clone)]
pub struct SpanInput {
    pub problem: String,
    pub answer:  String,
    pub reasoning: String,
}

#[derive(Debug, Clone)]
pub struct ProcessResult {
    pub mode: Mode,
    pub original: String,
    pub processed: String,
    pub original_tokens: usize,
    pub processed_tokens: usize,
}

pub fn format_compress_user(s: &SpanInput) -> String {
    format!(
        "PROBLEM:\n{}\n\nANSWER:\n{}\n\nREASONING:\n{}",
        if s.problem.trim().is_empty() { "(no surrounding problem context provided)" } else { s.problem.trim() },
        if s.answer.trim().is_empty()  { "(no final answer provided)" }              else { s.answer.trim()  },
        s.reasoning,
    )
}

pub fn format_invert_user(s: &SpanInput) -> String {
    format!(
        "Problem:\n{}\n\nModel's final answer:\n{}\n\nReasoning Bubbles:\n{}\n\nReconstruct the full reasoning trace.",
        if s.problem.trim().is_empty() { "(no surrounding problem context provided)" } else { s.problem.trim() },
        if s.answer.trim().is_empty()  { "(no final answer provided)" }              else { s.answer.trim()  },
        s.reasoning,
    )
}

pub struct Processor { pub mode: Mode, pub level: Level }

impl Processor {
    pub fn new(mode: Mode, level: Level) -> Self { Self { mode, level } }

    /// Local (no-LLM) processing for rules/noop modes.
    pub fn process_local(&self, s: &SpanInput) -> ProcessResult {
        let text = &s.reasoning;
        if text.trim().is_empty() {
            return ProcessResult {
                mode: self.mode, original: text.clone(), processed: text.clone(),
                original_tokens: 0, processed_tokens: 0,
            };
        }
        match self.mode {
            Mode::Noop => ProcessResult {
                mode: self.mode, original: text.clone(), processed: text.clone(),
                original_tokens: estimate_tokens(text), processed_tokens: estimate_tokens(text),
            },
            Mode::Rules => {
                let r = compress_rules(text, self.level);
                ProcessResult {
                    mode: self.mode,
                    original: r.original, processed: r.compressed,
                    original_tokens: r.original_tokens, processed_tokens: r.compressed_tokens,
                }
            }
            _ => panic!("process_local called for LLM mode"),
        }
    }

    pub fn system(&self) -> &'static str {
        match self.mode {
            Mode::Compress => prompts::compress_prompt(self.level),
            Mode::Invert   => prompts::invert_prompt(),
            _              => "",
        }
    }

    pub fn format_user(&self, s: &SpanInput) -> String {
        match self.mode {
            Mode::Compress => format_compress_user(s),
            Mode::Invert   => format_invert_user(s),
            _ => s.reasoning.clone(),
        }
    }

    /// Process a batch of spans concurrently against an LLM endpoint.
    pub async fn process_batch_llm(
        &self,
        client: LlmClient,
        spans: &[SpanInput],
        max_workers: usize,
    ) -> Vec<ProcessResult> {
        let system = self.system().to_string();
        let items: Vec<(String, String)> = spans.iter()
            .map(|s| (system.clone(), self.format_user(s))).collect();
        let outs = client.batch_chat(items, max_workers).await;

        spans.iter().zip(outs).map(|(span, out)| {
            let text = &span.reasoning;
            let processed = match out {
                Some(s) if !s.trim().is_empty() => s,
                _ => text.clone(), // fall back to original on permanent failure
            };
            ProcessResult {
                mode: self.mode,
                original: text.clone(),
                original_tokens: estimate_tokens(text),
                processed_tokens: estimate_tokens(&processed),
                processed,
            }
        }).collect()
    }
}
